apps/src/blockly/addons/cdoFieldNumber.ts

```diff
@@ -0,0 +1,162 @@
+import GoogleBlockly, {
+  FieldNumberConfig,
+  FieldNumberValidator,
+} from 'blockly/core';
+
+import {
+  ExtendedBlockSvg,
+  ExtendedConnection,
+  FieldHelperOptions,
+} from '../types';
+
+import CdoAngleHelper from './cdoAngleHelper';
+
+export default class CdoFieldNumber extends GoogleBlockly.FieldNumber {
+  private config: string | null | undefined;
+  angleHelper: CdoAngleHelper | null;
+  direction: string | undefined;
+  directionFieldName: string | undefined;
+
+  /**
+   * @param value The initial value of the field. Should cast to a number.
+   *     Defaults to 0. Also accepts Field.SKIP_SETUP if you wish to skip setup
+   *     (only used by subclasses that want to handle configuration and setting
+   *     the field value after their own constructors have run).
+   * @param min Minimum value. Will only be used if config is not
+   *     provided.
+   * @param max Maximum value. Will only be used if config is not
+   *     provided.
+   * @param precision Precision for value. Will only be used if config
+   *     is not provided.
+   * @param validator A function that is called to validate changes to the
+   *     field's value. Takes in a number & returns a validated number, or null
+   *     to abort the change.
+   * @param config A map of options used to configure the field.
+   *     See the [field creation documentation]{@link
+   * https://developers.google.com/blockly/guides/create-custom-blocks/fields/built-in-fields/number#creation}
+   * for a list of properties this parameter supports.
+   */
+  constructor(
+    value?: string | number,
+    min?: string | number | null,
+    max?: string | number | null,
+    precision?: string | number | null,
+    validator?: FieldNumberValidator | null,
+    config?: FieldNumberConfig
+  ) {
+    super(value, min, max, precision, validator, config);
+    this.angleHelper = null;
+  }
+
+  /**
+   * If this field is attached to a block whose output connection is attached to a
+   * connection that has the specified field helper, get the options for that
+   * field helper.
+   * @param {string} fieldHelper - the field helper to retrieve. One of
+   *        Blockly.BlockFieldHelper
+   * @return {Object|undefined} the options object if it exists
+   */
+  getFieldHelperOptions(fieldHelper: string) {
+    return (this.sourceBlock_ &&
+      this.sourceBlock_.outputConnection &&
+      this.sourceBlock_.outputConnection.targetConnection &&
+      (
+        this.sourceBlock_.outputConnection
+          .targetConnection as ExtendedConnection
+      ).getFieldHelperOptions(fieldHelper)) as FieldHelperOptions | undefined;
+  }
+
+  shouldShowAngleHelper() {
+    return this.getFieldHelperOptions(Blockly.BlockFieldHelper.ANGLE_HELPER);
+  }
+
+  protected showEditor_(e?: Event, quietInput?: boolean): void {
+    super.showEditor_(e, quietInput);
+    if (this.shouldShowAngleHelper()) {
+      this.initializeAngleHelper();
+    }
+  }
+
+  /**
+   * Initialize the angle helper.
```

Comment:
nit: explain what the angle helper is, and that we only use it if a certain field helper is specified on the parent block?